### PR TITLE
Converted time results from HEK Client to Astropy Time

### PR DESCRIPTION
**Pull request by jeffreypaul15**
### Description

Ref #5006
The `_convert_to_Time()` method replaces the `str` with the appropriate `astropy.time.core.Time` object.
The conversion to `astropy.units.quantity.Quantity` is still yet to be done based on input.

### CODE REVIEW

**Comment by Cadair** [comment]
Thanks for this PR, but following on from the conversation at the community meeting today I think that we shouldn't take this approach of converting known columns, we should work with the HEK's VO table or other mechanisms to solve this in a general manner.

The only other thing I would suggest is turning this into a helper function which can optionally be applied to the results to convert the known columns.

If you don't want to turn this PR into the latter option, then please close this.

**Comment by jeffreypaul15** [comment]
> Thanks for this PR, but following on from the conversation at the community meeting today I think that we shouldn't take this approach of converting known columns, we should work with the HEK's VO table or other mechanisms to solve this in a general manner.
> 
> The only other thing I would suggest is turning this into a helper function which can optionally be applied to the results to convert the known columns.
> 
> If you don't want to turn this PR into the latter option, then please close this.

Okay, that makes sense. What would the helper function contain though? How would we want to write it?

**Comment by nabobalis** [comment]
The helper function would probably look like the function you wrote, we would probably not call it automatically but tell users in a guide or docstring it's there for use.

**Comment by jeffreypaul15** [comment]
> The helper function would probably look like the function you wrote, we would probably not call it automatically but tell users in a guide or docstring it's there for use.

So, just convert this into a method that would be called only if the user desires right? How would we want to design that? Have a separate argument for this within the HEK client?

**Comment by nabobalis** [comment]
> > The helper function would probably look like the function you wrote, we would probably not call it automatically but tell users in a guide or docstring it's there for use.
> 
> So, just convert this into a method that would be called only if the user desires right?

Yeah.

> How would we want to design that? Have a separate argument for this within the HEK client?

I don't think this would be an argument, maybe a function a user can import separately and then apply on the results object.

I am unsure of the path required in code to make that an argument.

**Comment by jeffreypaul15** [comment]
> maybe a function a user can import separately and then apply on the results object.

Yeah this makes more sense.

**Comment by jeffreypaul15** [comment]
> > > The helper function would probably look like the function you wrote, we would probably not call it automatically but tell users in a guide or docstring it's there for use.
> > 
> > 
> > So, just convert this into a method that would be called only if the user desires right?
> 
> Yeah.
> 
> > How would we want to design that? Have a separate argument for this within the HEK client?
> 
> I don't think this would be an argument, maybe a function a user can import separately and then apply on the results object.
> 
> I am unsure of the path required in code to make that an argument.

Would this be exclusive to HEK Client responses? Or should I place this method elsewhere?

**Comment by nabobalis** [comment]
> Would this be exclusive to HEK Client responses? Or should I place this method elsewhere?

I don't see why it wouldn't be.

**Comment by jeffreypaul15** [comment]
> > Would this be exclusive to HEK Client responses? Or should I place this method elsewhere?
> 
> I don't see why it wouldn't be.

I'm a bit confused on where to place this method exactly, if it's not exclusive then do you have any suggestions?

**Comment by nabobalis** [comment]
> > > Would this be exclusive to HEK Client responses? Or should I place this method elsewhere?
> > 
> > 
> > I don't see why it wouldn't be.
> 
> I'm a bit confused on where to place this method exactly, if it's not exclusive then do you have any suggestions?

Same place as the client?

**Comment by jeffreypaul15** [comment]
> > > > Would this be exclusive to HEK Client responses? Or should I place this method elsewhere?
> > > 
> > > 
> > > I don't see why it wouldn't be.
> > 
> > 
> > I'm a bit confused on where to place this method exactly, if it's not exclusive then do you have any suggestions?
> 
> Same place as the client?

But we want this to work for all clients right?

**Comment by nabobalis** [comment]
> > > > > Would this be exclusive to HEK Client responses? Or should I place this method elsewhere?
> > > > 
> > > > 
> > > > I don't see why it wouldn't be.
> > > 
> > > 
> > > I'm a bit confused on where to place this method exactly, if it's not exclusive then do you have any suggestions?
> > 
> > 
> > Same place as the client?
> 
> But we want this to work for all clients right?

No?

**Comment by jeffreypaul15** [comment]
> > > > > > Would this be exclusive to HEK Client responses? Or should I place this method elsewhere?
> > > > > 
> > > > > 
> > > > > I don't see why it wouldn't be.
> > > > 
> > > > 
> > > > I'm a bit confused on where to place this method exactly, if it's not exclusive then do you have any suggestions?
> > > 
> > > 
> > > Same place as the client?
> > 
> > 
> > But we want this to work for all clients right?
> 
> No?

Oh okay, I misunderstood. Sorry.

**Comment by jeffreypaul15** [comment]
```python
from sunpy.net import Fido
from sunpy.net import attrs as a
from sunpy.net.hek.hek import convert_time
import sunpy
from astropy.time import Time

event_type = "FL"
tstart = "2013/10/28"
tend = "2013/10/29"
result = Fido.search(a.Time(tstart, tend),
                     a.hek.EventType(event_type),
                     a.hek.OBS.Observatory == "GOES")

convert_time(result)

```
This should produce the required result.
I do have a few questions though :
- Would it be better to ` __init__` this method so the user imports it from from `sunpy.net.hek` instead of from `sunpy.net.hek.hek`?
- I'm importing `from astropy.time import Time`, should I just use `sunpy.time.parse_time` instead of having that import?
- Does this need a test?

**Comment by nabobalis** [comment]
> * Would it be better to ` __init__` this method so the user imports it from from `sunpy.net.hek` instead of from `sunpy.net.hek.hek`?

Sure.

> * I'm importing `from astropy.time import Time`, should I just use `sunpy.time.parse_time` instead of having that import?

Yes.

> * Does this need a test?

Yes.

Ideally the function would do more, but it's ok for as a starting platform.

**Comment by jeffreypaul15** [comment]
@nabobalis is this good to go?

File: changelog/5131.feature.rst [hunk 1]
```diff
@@ -0,0 +1,2 @@
+Added a :meth:`~sunpy.net.hek._convert_to_Time` method to `sunpy.net.hek`
```
Comment (nabobalis) [inline changelog/5131.feature.rst:1]:
```suggestion
Added a :meth:`~sunpy.net.hek.convert_time` method to `sunpy.net.hek`
```

File: sunpy/net/hek/__init__.py [hunk 2]
```diff
@@ -1 +1,2 @@
 from sunpy.net.hek.hek import *
+from sunpy.net.hek.hek import convert_time
```
Comment (nabobalis) [inline sunpy/net/hek/__init__.py:2]:
```suggestion
```

File: sunpy/net/hek/hek.py [hunk 3]
```diff
@@ -9,13 +9,13 @@
...
-from astropy.time import Time
 
 import sunpy.net._attrs as core_attrs
 from sunpy import log
 from sunpy.net import attr
 from sunpy.net.base_client import BaseClient, QueryResponseTable
 from sunpy.net.hek import attrs
+from sunpy.time import parse_time
 from sunpy.util import dict_keys_same, unique
 from sunpy.util.xml import xml_to_dict
```
Comment (nabobalis) [inline sunpy/net/hek/hek.py:21]: Add the function to the `__all__` below this.

File: sunpy/net/hek/hek.py [hunk 4]
```diff
@@ -33,6 +33,17 @@ def _freeze(obj):
     return obj
 
 
+def convert_time(response_obj):
+    """ Helper method that convert time from str to astropy.time.core.Time"""
```
Comment (nabobalis) [inline sunpy/net/hek/hek.py:37]:
```suggestion
    """
    Converts time from str to astropy.time.core.Time for a hek response.
    """
```

File: sunpy/net/hek/tests/test_hek.py [hunk 5]
```diff
@@ -197,3 +199,12 @@ def test_mixed_results_get_angstrom():
     result = client.search(attrs.Time(tstart, tend), attrs.hek.EventType(event_type))
     assert len(result) == 13
     assert result[0]["SOL_standard"] == 'SOL2014-10-24T20:53:46L247C106'
+
+
+@pytest.mark.remote_data
+def test_convert_time(hek_client_creator):
```
Comment (nabobalis) [inline sunpy/net/hek/tests/test_hek.py:205]: Does it work with a return from fido?

Comment (jeffreypaul15) [reply sunpy/net/hek/tests/test_hek.py:205]: Yes, it does. Would we need a test for that as well?

Comment (nabobalis) [reply sunpy/net/hek/tests/test_hek.py:205]: Sure.

Comment (jeffreypaul15) [reply sunpy/net/hek/tests/test_hek.py:205]: > Sure.

Where would this test go though? Should I import fido and write it here itself?

Comment (nabobalis) [reply sunpy/net/hek/tests/test_hek.py:205]: Yeah.

File: sunpy/net/hek/hek.py [hunk 6]
```diff
@@ -33,6 +33,19 @@ def _freeze(obj):
     return obj
 
 
+def convert_time(response_obj):
+    """
+    Converts time from str to astropy.time.core.Time for a HEKTable.
+    """
```
Comment (nabobalis) [inline sunpy/net/hek/hek.py:39]: I think this docstring needs expanding to have a parameters section, also a note to say this does the conversion in place. 

Maybe we want it to return a new repesonse?

**Comment by dstansby** [comment]
I'm -1 to merging something like this that isn't a generic solution, as it just increases the amount of code we have to maintain without *really* solving the issue in the long run (as it stands we're just guessing at column names that need converting).

**Comment by dstansby** [comment]
I think it's worth discussing this specific PR at our next weekly meeting, and we can come to a collective decision on whether to go forward with it or not.

**Comment by nabobalis** [comment]
Following up in this weekly meeting. We don't want to add code to the client that deals with conversion of the table columns in an ad hoc manner. We need to work out a schema for the conversion to start with.

**Pull request closed without merging**